### PR TITLE
[Mobile Payments] Modal views

### LINKED ISSUE #3980: [Mobile Payments] [UI] Modal view
For context: p91TBi-4KR-p2 and p91TBi-4Af-p2

We need to notify users of the status of the payment collection process: notify when they need to tap the card, when they need to remove the card, potential errors...

We need to do that in a modal view that supports updating its content without having to be dismissed and presented again.

First, we need to check if the current FancyAlertViewController would support that, and if it doesn't, we'd need to roll our own

### PR DESCRIPTION
**Pull request by ctarda**
Closes #3980 

⚠️ This PR is against the feature branch implementing support for Card Present Payments, not against develop ⚠️
⚠️ Testing this PR requires a hardware reader ⚠️

This is a first iteration on the final layout of the modal views. I don't love this implementation, but we need to push something forward now in order to unblock our efforts to provide a better UI for discovering and pairing with cards.

## Changes
* Add a view controller that contains all the labels and buttons used by the different states of the modal view that are necessary to collect a payment. This might not be the best possible approach, but it is what we need at this point.
* This is responsible for styling and showing/hiding labels and buttons.
* Add different "viewmodels" that provide the content for the view controller to render, and encapsulate the actions associated to the buttons.

Again, this is not perfect, but it's better than nothing, and it moves the needle forward. We can start logging tickets for improvements right away. 😄 

https://user-images.githubusercontent.com/2722505/116465776-acc41300-a83b-11eb-9982-87558c475c7b.mov

| Light appearance |  Dark Appearance |
| ---- | ---- |
|  |  |
|  |  |
|  |  |

## How to test
Testing requires:

1. Setting up a store for Card Present Payment Testing (also P91TBi-4BH-p2 for more specific instructions)
2. An external hardware reader.

* Checkout the branch, run bundle exec pod install, just for kicks and giggles.
* Build and run on a device.
* Switch an external card reader on. The only card reader supported by now is the BBPOS Chipper 2X BT.
* Log in to an account that matches a store that has been configured for testing.
* Navigate to Settings in the app. (Tap the gear button in the top right)
* In the "Store Settings" section, tap "Manage Card Reader".
* Make sure the card reader is on and the blue light is blinking (we don't do any error handling yet, so it's important that things are setup for the happy path)
* Tap "Connect card reader"
* Wait until the app suggests one reader to connect to.
* Tap "Connect to reader"
* Make sure the reader light turns solid blue.
* Navigate to an order eligible for Card Present Payments (it has to be a store enrolled to WCPay, and the order needs to have as payment method associated Cash on Delivery or none).
* Tap the "Collect Payment" button.
* Tap the card to the reader when requested.
* If everything goes well, the card should be charged, and the payment should be completed.
* At that point, tapping the "Print receipt" button will trigger a printout. The "Email receipt button" does nothing yet.

Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: Hardware/Hardware/CardReader/CardReaderService.swift [hunk 1]
```diff
@@ -49,10 +49,6 @@ public protocol CardReaderService {
     /// Captures a payment after collecting a payment method succeeds.
     func capturePayment(_ parameters: PaymentIntentParameters) -> AnyPublisher<PaymentIntent, Error>
 
-    /// Cancels a a PaymentIntent
```
Comment (allendav) [inline Hardware/Hardware/CardReader/CardReaderService.swift]: Why this change?

File: WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentModalError.swift [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+final class CardPresentModalError: CardPresentPaymentsModalViewModel {
+
+    /// Aount charged
+    private let amount: String
+
+    /// The error returned by the stack
+    private let error: Error
+
+    /// A closure to ececute when the primary button is tapped
+    private let tryAgainAction: () -> Void
```
Comment (allendav) [inline WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentModalError.swift]: Would this be better named as `primaryAction` ?

File: WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentModalError.swift [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+/// Modal presented on error
+final class CardPresentModalError: CardPresentPaymentsModalViewModel {
+
+    /// Aount charged
+    private let amount: String
+
+    /// The error returned by the stack
+    private let error: Error
+
+    /// A closure to ececute when the primary button is tapped
```
Comment (allendav) [inline WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentModalError.swift]: execute

File: WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentModalError.swift [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+    let secondaryButtonTitle: String? = nil
+
+    var bottomTitle: String? {
+        error.localizedDescription
+    }
+
+    let bottomSubtitle: String? = nil
+
+    init(amount: String, error: Error, tryAgain: @escaping () -> Void) {
```
Comment (allendav) [inline WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentModalError.swift]: Could we name this `primaryAction` or `didTapPrimaryButton` or something? Or will all error paths always be re-tryable?

File: WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentPaymentsModalViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    var bottomSubtitle: String? { get }
+
+    /// Executes action associated to a tap in the view controller primary button
+    /// - Parameter viewController: usually the view controller sending the tap
+    func didTapPrimaryButton(in viewController: UIViewController?)
+
+    /// Executes action associated to a tap in the view controller secondary button
+    /// - Parameter viewController: usually the view controller sending the tap
+    func didTapSecondaryButton(in viewController: UIViewController?)
+}
```
Comment (allendav) [inline WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentPaymentsModalViewModel.swift:37]: Why isn't the aux button part of the protocol too?

File: WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentModalError.swift [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+    func didTapPrimaryButton(in viewController: UIViewController?) {
+        tryAgainAction()
+        viewController?.dismiss(animated: true)
+    }
+
+    func didTapSecondaryButton(in viewController: UIViewController?) {
+    }
+
+    func didTapAuxiliaryButton(in viewController: UIViewController?) {
```
Comment (allendav) [inline WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentModalError.swift]: See question about protocol below

File: WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentModalRemoveCard.swift [hunk 7]
```diff
@@ -0,0 +1,58 @@
+import UIKit
+
+
+/// Modal presented when the card reader requests customers to remove the card
+final class CardPresentModalRemoveCard: CardPresentPaymentsModalViewModel {
+
+    /// Customer name
+    private let name: String
+
+    /// Charge ampnt
```
Comment (allendav) [inline WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentModalRemoveCard.swift]: amount

File: WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentModalSuccess.swift [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    /// Amount charged
+    private let amount: String
+
+    /// Closure to execute when primary button is tapped
+    private let printReceiptAction: () -> Void
+
+
+    /// Closure to execute when secondary button is tapped
+    private let emailReceiptAction: () -> Void
+
```
Comment (allendav) [inline WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentModalSuccess.swift:15]: Did you forget to hook up noThanks?

File: WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentPaymentsModalViewModel.swift [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    case fullInfo
+
+    /// From top to bottom: Two lines of text at the top, image, one more line of text
+    case reducedInfo
+
+    /// From top to bottom: Two lines of text at the top, image, two action buttons
+    case twoActionButtons
+
+    /// From top to bottom: Two lines of text at the top, image, one action button
+    case oneActionButton
```
Comment (allendav) [inline WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentPaymentsModalViewModel.swift:52]: Needs one more case for aux, right?

File: WooCommerce/Classes/ViewModels/Order Details/OrderDetailsViewModel.swift [hunk 10]
```diff
@@ -503,6 +503,7 @@ extension OrderDetailsViewModel {
     }
 
     func printReceipt(params: CardPresentReceiptParameters) {
+        print("==== reaching print receipt in viewmodel")
```
Comment (allendav) [inline WooCommerce/Classes/ViewModels/Order Details/OrderDetailsViewModel.swift]: Still needed?

File: WooCommerce/Classes/ViewRelated/CardPresentPayments/CardPresentPaymentsModalViewController.swift [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+        populateContent()
+    }
+
+    func setViewModel(_ newViewModel: CardPresentPaymentsModalViewModel) {
+        self.viewModel = newViewModel
+
+        populateContent()
+    }
+
+    override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
```
Comment (allendav) [inline WooCommerce/Classes/ViewRelated/CardPresentPayments/CardPresentPaymentsModalViewController.swift]: Why is this needed? And viewWillTransition also?

File: WooCommerce/WooCommerceTests/ViewRelated/CardPresentPayments/CardPresentPaymentsModalViewControllerTests.swift [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+    func test_viewcontroller_presents_bottom_subtitle_provided_by_viewmodel() throws {
+        let viewModel = ModalViewModel(mode: .fullInfo)
+        let viewController = CardPresentPaymentsModalViewController(viewModel: viewModel)
+
+        _ = try XCTUnwrap(viewController.view)
+
+        XCTAssertEqual(viewController.getBottomSubtitleLabel().text, viewModel.bottomSubtitle)
+    }
+
+    func test_viewcontroller_propagates_tap_in_primary_button_to_viewmodel() throws {
```
Comment (allendav) [inline WooCommerce/WooCommerceTests/ViewRelated/CardPresentPayments/CardPresentPaymentsModalViewControllerTests.swift:52]: For more coverage, you may want to test primary button against .oneActionButton and .reducedInfoOneActionButton too

File: WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentPaymentsModalViewModel.swift [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+
+    /// From top to bottom: One line of text at the top, image, one action button
+    case reducedInfoOneActionButton
```
Comment (allendav) [inline WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentPaymentsModalViewModel.swift:55]: This enum suggests we might want two enums to describe the modal's content - one for info level and the other for the number of buttons.

File: WooCommerce/Classes/ViewModels/Order Details/OrderDetailsPaymentAlerts.swift [hunk 14]
```diff
@@ -8,47 +8,69 @@ import WordPressUI
 /// to rewrite whenever we have the UI finalized.
 /// https://github.com/woocommerce/woocommerce-ios/issues/3980
 final class OrderDetailsPaymentAlerts {
-    private var alertController: FancyAlertViewController?
-    private var name: String?
-    private var amount: String?
+    //private var alertController: FancyAlertViewController?
```
Comment (allendav) [inline WooCommerce/Classes/ViewModels/Order Details/OrderDetailsPaymentAlerts.swift]: Did you want to go ahead and delete this now?

**Review by allendav — APPROVED** [review]
A few suggestions. Tests well. CardPresentPaymentsModalViewControllerTests also pass. :shipit:

File: Hardware/Hardware/CardReader/CardReaderService.swift (hunk 1 above)
Comment (ctarda) [reply Hardware/Hardware/CardReader/CardReaderService.swift]: We don't really need it at this point, it looks like I was a bit trigger happy with the delete button. It shouldn't have made it to this PR, I'll undo and log a separate ticket for cleaning up the service

Comment (ctarda) [reply Hardware/Hardware/CardReader/CardReaderService.swift]: #4068

File: WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentPaymentsModalViewModel.swift (hunk 5 above)
Comment (ctarda) [reply WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentPaymentsModalViewModel.swift:37]: It's an oversight on my side. 😊 I logged #4069 to add it as a follow up

File: WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentModalError.swift (hunk 6 above)
Comment (ctarda) [reply WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentModalError.swift]: I've removed all mentions to the auxiliaryButton from this PR. I'll come back to it in #4069

File: WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentModalSuccess.swift (hunk 8 above)
Comment (ctarda) [reply WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentModalSuccess.swift:15]: Yep, that would be the auxiliary button, I guess. Will add in #4069

File: WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentPaymentsModalViewModel.swift (hunk 9 above)
Comment (ctarda) [reply WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentPaymentsModalViewModel.swift:52]: Correct. #4069

File: WooCommerce/Classes/ViewRelated/CardPresentPayments/CardPresentPaymentsModalViewController.swift (hunk 11 above)
Comment (ctarda) [reply WooCommerce/Classes/ViewRelated/CardPresentPayments/CardPresentPaymentsModalViewController.swift]: My bad, I will remove them. The reason I added them is because we will need this to support dynamic type and rotation properly (#4064)

* when rotating the device, we'll need to change the main stack view's axis from vertical to horizontal
* for certain font sizes we might want to tweak certain properties in the stack view.

We can do those things overriding this method here.

File: WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentPaymentsModalViewModel.swift (hunk 13 above)
Comment (ctarda) [reply WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentPaymentsModalViewModel.swift:55]: Good point. Let's split it as soon as we add a new case

**Comment by ctarda** [comment]
Thanks for the review @allendav I just pushed an update to incorporate your feedback, with a couple of exceptions that I have deferred to a subsequent PR. I'll merge as soon as CI is done
